spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/DataFrameImplicits.scala

```diff
@@ -83,6 +85,68 @@ object DataFrameImplicits {
       }
     }
 
+    /**
+     * Casts NullType (aka VOID) fields to their target types as defined in `targetSchema`.
+     *
+     * Matching of fields is "by name", so order of fields in schema(s) doesn't matter.
+     * Resulting DataFrame has the same fields order as original DataFrame.
+     *
+     * @param targetSchema definition of field types to which potential NullTypes will be casted to
+     * @return DataFrame with fields of NullType casted to their type in `targetSchema`
+     */
+    def enforceTypeOnNullTypeFields(targetSchema: StructType): DataFrame = {
+
+      def processArray(
+                        thisArrType: ArrayType, targetArrType: ArrayType, thisArrayColumn: Column
+                      ): Column =
+        (thisArrType.elementType, targetArrType.elementType) match {
+          case (_: NullType, _) =>
+            transform(
+              thisArrayColumn,
+              _ => lit(null).cast(targetArrType.elementType)
+            )
+          case (thisNestedArrType: ArrayType, targetNestedArrType: ArrayType) =>
+            transform(
+              thisArrayColumn,
+              processArray(thisNestedArrType, targetNestedArrType, _)
+            )
+          case (thisNestedStructType: StructType, targetNestedStructType: StructType) =>
+            transform(
+              thisArrayColumn,
+              element => struct(processStruct(thisNestedStructType, targetNestedStructType, Some(element)): _*)
+            )
+          case _ => thisArrayColumn
+        }
+
+      def processStruct(
+                         currentThisSchema: StructType, currentTargetSchema: StructType, parent: Option[Column]
+                       ): List[Column] = {
+        val currentTargetSchemaMap = currentTargetSchema.map(f => (f.name.toLowerCase, f)).toMap
+
+        currentThisSchema.foldRight(List.empty[Column])((field, acc) => {
+          val currentColumn: Column = parent
+            .map(_.getField(field.name))
+            .getOrElse(col(field.name))
+            .as(field.name)
+          val correspondingTargetType = currentTargetSchemaMap.get(field.name.toLowerCase).map(_.dataType)
+
+          (field.dataType, correspondingTargetType) match {
+            case (NullType, Some(targetType)) =>
+              currentColumn.cast(targetType).as(field.name) :: acc
+            case (arrType: ArrayType, Some(targetArrType: ArrayType)) =>
+              processArray(arrType, targetArrType, currentColumn).as(field.name) :: acc
+            case (structType: StructType, Some(targetStructType: StructType)) =>
+              struct(processStruct(structType, targetStructType, Some(currentColumn)): _*).as(field.name) :: acc
+            case _ => currentColumn :: acc
```

Comment:
**Possibly major:**
This will result the output of the function will contain even fields that are not present in the target schema. I think that is not desired by the expected logic of the function. 🤔

Reply:
IMO it is desired - all the function does is enforcing types on NullType, ofc in case if field is present in `targetSchema` (otherwise it cannot enforce anything). Added a line in docstring saying what is the desired behaviour.

Reply:
OK, that makes sense, and the added _docstring_ helps considerably in this regard. 😎 

In that case I would suggest, to avoid code repetition, to assign the result of the match-case statement to a `val` of type column, eg. `val castedColumn`. And only then construct the result:
`castedColumn.as(field.name) :: acc`

Reply:
Good idea again, done